File: node/checks/retries.ts
/**
 * Retry checks - 429s, Retry-After, backoff
 */

import fetch from 'node-fetch';
import { Config } from '../config';

interface CheckResult {
  status: 'pass' | 'warn' | 'fail' | 'skipped';
  findings: Array<{
    severity: 'info' | 'warning' | 'error';
    message: string;
    details?: any;
  }>;
  metrics?: Record<string, any>;
}

export async function checkRetries(config: Config): Promise<CheckResult> {
  const findings: any[] = [];
  const metrics: Record<string, any> = {};

  try {
    const url = `${config.baseUrl.replace(/\/$/, '')}/v1/chat/completions`;
    const headers = {
      Authorization: `Bearer ${config.apiKey}`,
      'Content-Type': 'application/json',
    };

    const payload = {
      model: config.model || 'gpt-3.5-turbo',
      messages: [{ role: 'user', content: 'Test' }],
      max_tokens: 10,
    };

    const response = await fetch(url, {
      method: 'POST',
      headers,
      body: JSON.stringify(payload),
    });

    // Check for rate limit headers
    const retryAfter = response.headers.get('retry-after');
    if (retryAfter) {
      findings.push({
        severity: 'info',
        message: `Retry-After header present: ${retryAfter}s`,
      });
      metrics.retry_after_s = retryAfter;
    }

    const rateLimitRemaining = response.headers.get('x-ratelimit-remaining');
    if (rateLimitRemaining) {
      metrics.ratelimit_remaining = rateLimitRemaining;

      if (parseInt(rateLimitRemaining) < 10) {
        findings.push({
          severity: 'warning',
          message: `Low rate limit remaining: ${rateLimitRemaining}`,
        });
      }
    }

    // General recommendations
    findings.push({
      severity: 'info',
      message: 'Recommended: Use exponential backoff with jitter for retries',
    });

    findings.push({
      severity: 'info',
      message: 'Never retry after stream has started (partial response received)',
    });

    findings.push({
      severity: 'info',
      message: 'Set retry cap (e.g., 3 attempts max) to avoid infinite loops',
    });

    return {
      status: 'pass',
      findings,
      metrics,
    };
  } catch (error: any) {
    if (error.message.includes('429')) {
      return {
        status: 'warn',
        findings: [
          {
            severity: 'warning',
            message: 'Rate limited (429). Check Retry-After header.',
          },
        ],
        metrics,
      };
    }

    return {
      status: 'fail',
      findings: [
        {
          severity: 'error',
          message: `Retry check failed: ${error.message}`,
        },
      ],
      metrics,
    };
  }
}

File: python/src/ai_patch/checks/retries.py
"""Retry checks - 429s, Retry-After, backoff."""

import httpx
from typing import Dict, Any
from ai_patch.config import Config


def check(config: Config) -> Dict[str, Any]:
    """Run retry checks."""
    
    findings = []
    metrics = {}
    
    try:
        # Test for 429 handling
        url = f"{config.base_url.rstrip('/')}/v1/chat/completions"
        headers = {
            "Authorization": f"Bearer {config.api_key}",
            "Content-Type": "application/json"
        }
        
        payload = {
            "model": config.model or "gpt-3.5-turbo",
            "messages": [{"role": "user", "content": "Test"}],
            "max_tokens": 10
        }
        
        # Make a test request
        response = httpx.post(url, headers=headers, json=payload, timeout=30.0)
        
        # Check for rate limit headers
        if 'retry-after' in response.headers:
            retry_after = response.headers['retry-after']
            findings.append({
                'severity': 'info',
                'message': f'Retry-After header: {retry_after}s'
            })
            metrics['retry_after_s'] = retry_after
        
        if 'x-ratelimit-remaining' in response.headers:
            remaining = response.headers['x-ratelimit-remaining']
            metrics['ratelimit_remaining'] = remaining
            
            if int(remaining) < 10:
                findings.append({
                    'severity': 'warning',
                    'message': f'Rate limit remaining: {remaining} requests'
                })
        
        # Check for 429 status
        if response.status_code == 429:
            findings.append({
                'severity': 'warning',
                'message': 'Rate limiting detected (HTTP 429)'
            })
        
        status = 'warn' if any(f['severity'] in ['warning', 'error'] for f in findings) else 'pass'
        
    except httpx.HTTPStatusError as e:
        if e.response.status_code == 429:
            status = 'warn'
            retry_after = e.response.headers.get('retry-after', 'not set')
            findings.append({
                'severity': 'warning',
                'message': f'Rate limited (429). Retry-After: {retry_after}'
            })
        else:
            status = 'fail'
            findings.append({
                'severity': 'error',
                'message': f'HTTP error {e.response.status_code}'
            })
    except Exception as e:
        status = 'fail'
        findings.append({
            'severity': 'error',
            'message': f'Retry check failed: {str(e)}'
        })
    
    return {
        'status': status,
        'findings': findings,
        'metrics': metrics
    }
